### PR TITLE
[LangRef] Try to clarify some Metadata semantics

### PR DESCRIPTION
**Pull request by slinder1**
General cleanup in LangRef (and two outdated comments in LLParser.cpp) with the aim of making it easier to understand some of the terminology and subtle idiosyncrasies related to metadata in the IR.

I'm still not happy with the fact that "node" is used both informally and with a particular category of metadata in mind, depending on the context. This also bleeds into the type names in the implementation.

There are also several places where names from the implementation appear in the document with no other context or definition. In some cases I added a parenthetical to section titles to tie the two together, but I don't think this is ideal.

I also think it might be useful to define the "abstract" metadata classes like "DIScope" in the document, so the hierarchy of metadata node kinds is direct, and so we can avoid repetitive descriptions of all of the members of on part of the hierarchy. This inheritance doesn't have to be in terms of C++ classes, but using the same names as the implementation seems helpful, and we already do it for many other things.

Finally I added sections for the specialized nodes which are implemented in the IR but didn't have documentation in LangRef yet. These could use some work, and I admit I didn't dig too deep into the specifics beyond enumerating the fields, but I think we would ideally always have a LangRef section for every kind of node.

### CODE REVIEW

File: llvm/docs/LangRef.rst [hunk 1]
```diff
@@ -5574,34 +5574,106 @@ occurs on.
...
-Metadata does not have a type, and is not a value. If referenced from a
-``call`` instruction, it uses the ``metadata`` type.
+There are two metadata primitives: strings and nodes. There are
+also specialized nodes which have a distinguished name and a set of named
+arguments.
+
+.. note::
+
+    One example application of metadata is source-level debug information,
+    which is currently the only use of specialized nodes.
```
Comment (adrian-prantl) [inline llvm/docs/LangRef.rst:5588]:
```suggestion
    which is currently the only user of specialized nodes.
```

**Review by adrian-prantl — APPROVED** [review]

**Review by jryans — APPROVED** [review]
Looks good to me, it's great to have these details recorded for future readers. 😄

File: llvm/docs/LangRef.rst [hunk 2]
```diff
@@ -5574,34 +5574,106 @@ occurs on.
...
 contain any character by escaping non-printable characters with
 "``\xx``" where "``xx``" is the two digit hex code. For example:
 "``!"test\00"``".
 
-Metadata nodes are represented with notation similar to structure
-constants (a comma separated list of elements, surrounded by braces and
-preceded by an exclamation point). Metadata nodes can have any values as
+.. note::
+
+   A metadata string is metadata, but is not a meatdata node.
```
Comment (rovka) [inline llvm/docs/LangRef.rst:5659]: Typo: meatdata

**Review by jmorse — APPROVED** [review]
LGTM. It's hard to pin down the details of `ValueAsMetadata` and `MetadataAsValue` because they don't follow principled rules (AFAIUI) and exist to glue the Value and Metadata hierarchies together at various points. It's the portion of debug-info that isn't a first-class element of LLVM, so it's no surprise it's hard to document.

File: llvm/docs/LangRef.rst (hunk 2 above)
Comment (slinder1) [reply llvm/docs/LangRef.rst]: Thank you!

**Pull request merged**